src/solc_0.7/common/Game/GameToken.sol

```diff
@@ -0,0 +1,381 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.7.1;
+pragma experimental ABIEncoderV2;
+
+import "../BaseWithStorage/ERC721BaseToken.sol";
+import "@openzeppelin/contracts/utils/EnumerableSet.sol";
+import "../Interfaces/AssetToken.sol";
+import "../Interfaces/GameToken.sol";
+import "../Libraries/SafeMathWithRequire.sol";
+
+// @review remove all console.logs !
+// import "hardhat/console.sol";
+
+contract GameToken is ERC721BaseToken, GameTokenInterface {
+    ///////////////////////////////  Libs //////////////////////////////
+
+    using EnumerableSet for EnumerableSet.UintSet;
+    using SafeMathWithRequire for uint256;
+
+    ///////////////////////////////  Data //////////////////////////////
+
+    struct Data {
+        EnumerableSet.UintSet _assets;
+        mapping(uint256 => uint256) _values;
+    }
+
+    uint256 public _nextId;
+
+    address internal _minter;
+    AssetToken _asset;
+
+    uint256 private constant CREATOR_OFFSET_MULTIPLIER = uint256(2)**(256 - 160);
+
+    mapping(uint256 => Data) private _gameData;
+    mapping(address => address) private _creatorship; // creatorship transfer
+
+    mapping(uint256 => string) private _metaData;
+    mapping(uint256 => mapping(address => bool)) private _gameEditors;
+
+    ///////////////////////////////  Events //////////////////////////////
+
+    event Minter(address newMinter);
+    event AssetsAdded(uint256 indexed id, uint256[] assets, uint256[] values);
+    event AssetsRemoved(uint256 indexed id, uint256[] assets, uint256[] values, address to);
+
+    constructor(
+        address metaTransactionContract,
+        address admin,
+        AssetToken asset
+    ) ERC721BaseToken(metaTransactionContract, admin) {
+        _asset = asset;
+        _nextId = 1;
+    }
+
+    ///////////////////////////////  Functions //////////////////////////////
+
+    function removeSingleAsset(
+        uint256 gameId,
+        uint256 assetId,
+        address to
+    ) external override {
+        require(msg.sender == _ownerOf(gameId) || _gameEditors[gameId][msg.sender], "ACCESS_DENIED");
+        require(to != address(0), "INVALID_TO_ADDRESS");
+        // "sub" is from SafeMathWithRequire.sol
+        _gameData[gameId]._values[assetId] = _gameData[gameId]._values[assetId].sub(1);
+        uint256 remainingAssets = _gameData[gameId]._values[assetId];
+
+        if (remainingAssets == 0) {
+            // "remove" is from EnumerableSet.sol
+            require(_gameData[gameId]._assets.remove(assetId), "ASSET_NOT_IN_GAME");
+        }
+
+        _asset.safeTransferFrom(address(this), to, assetId);
```

Comment:
I would use ERC1155 transfer function here else this would fails if you add 1 ERC1155 (if not erc721)

Reply:
yep, fixed.